### PR TITLE
Duplication of dynamic fsms

### PR DESCRIPTION
**Pull request by parthsarkar17**
Implements the option to create several (default 2) FSM registers that represent dynamic control flow, so that queries can be split between all of identical registers.

### CODE REVIEW

**Comment by rachitnigam** [comment]
@parthsarkar17 anything stopping this from being merged?

**Comment by parthsarkar17** [comment]
> @parthsarkar17 anything stopping this from being merged?

Yep, we wanted to benchmark the improvements. We found that duplicating FSM logic for each register was a bad idea, so I modified the code to share logic among the copied register. I think I can go ahead and make a full PR now, though.

File: calyx-opt/src/passes/top_down_compile_control.rs [hunk 1]
```diff
@@ -388,195 +475,178 @@ impl<'b, 'a> Schedule<'b, 'a> {
...
-                        assigns
-                    }
+                .flat_map(|(state, mut assigns)| {
+                    // find the register from which to query; try to split evenly among registers
+                    let (fsm, used_slicers) = {
+                        let num_registers: u64 =
+                            fsms.len().try_into().unwrap();
+                        let reg_to_query : usize = (state * num_registers
+                            / (fsm_rep.last_state)).try_into().unwrap();
+
```
Comment (calebmkim) [inline calyx-opt/src/passes/top_down_compile_control.rs:519]: Just double checking: this math is basically trying to find an even split among the regsiter queries?

File: calyx-opt/src/passes/top_down_compile_control.rs [hunk 2]
```diff
@@ -388,195 +475,178 @@ impl<'b, 'a> Schedule<'b, 'a> {
...
-                        fsm.borrow().get("in"),
-                        ec_borrow.get("out"),
-                        trans_guard.clone(),
-                    ),
-                    self.builder.build_assignment(
-                        fsm.borrow().get("write_en"),
-                        signal_on.borrow().get("out"),
-                        trans_guard,
+                // get a transition guard for the first fsm register, and apply it to every fsm register
+                let state_guard = Self::build_query(
```
Comment (calebmkim) [inline calyx-opt/src/passes/top_down_compile_control.rs:549]: Will picking the first register lead to uneven query-ing?

**Review by calebmkim — COMMENTED** [review]
This looks very good so far! 

One suggestion, if you think it's a viable idea: create a `DynamicFSM` struct (similar to the `StaticFSM` [struct](https://github.com/calyxir/calyx/blob/259c2c85991dc3eb1af118ec89a4aa453e71e7c5/calyx-opt/src/analysis/static_schedule.rs#L64)). 

This can have internal fields to hold the existing queries, the type of encoding (one-hot vs binary),  the actual FSM cell(s) (i.e., the `ir::RRC<ir::Cell>`), etc. 

I got this idea when I was looking at your code to evenly split FSM queries. I think it will probably be fine in practice, but I think it's slightly less-than-ideal for two reasons: 
1) Suppose you query from the FSM in a different part of the code. Each time you query from the FSM, you have to remember to do the math on the query to figure out which FSM should support the query (i.e., you would have to duplicate the code). 
2) Also, the math you're doing is more of a heuristic than an exact thing. Suppose, for whatever reasons, even numbered states were more heavily queried than odd number states. Then you might want to split the FSM queries differently. 

Lmk if this makes sense, and if it does, whether you think it would be worth implementing. It's just a thought on my end, so don't take it as something you have to do.

File: calyx-opt/src/passes/top_down_compile_control.rs (hunk 2 above)
Comment (parthsarkar17) [reply calyx-opt/src/passes/top_down_compile_control.rs:549]: As of now, there's only one duplicate register, so I think using the output of the first register's logic for the second register shouldn't be an issue. If we end up needing more duplicate registers, then that might create some bottleneck, but I think we should be good for now?

**Pull request merged**